src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -744,51 +842,68 @@ await ExecuteWithFailOverPolicyAsync(
-                var selector = new SettingSelector
-                {
-                    KeyFilter = KeyFilter.Any,
-                    LabelFilter = LabelFilter.Null
-                };
-
-                await CallWithRequestTracing(async () =>
+                if (string.IsNullOrEmpty(loadOption.SnapshotName))
                 {
-                    await foreach (ConfigurationSetting setting in client.GetConfigurationSettingsAsync(selector, cancellationToken).ConfigureAwait(false))
+                    var selector = new SettingSelector()
                     {
-                        serverData[setting.Key] = setting;
-                    }
-                }).ConfigureAwait(false);
-            }
+                        KeyFilter = loadOption.KeyFilter,
+                        LabelFilter = loadOption.LabelFilter
+                    };
 
-            foreach (KeyValueSelector loadOption in _options.KeyValueSelectors)
-            {
-                IAsyncEnumerable<ConfigurationSetting> settingsEnumerable;
+                    var matchConditions = new List<MatchConditions>();
 
-                if (string.IsNullOrEmpty(loadOption.SnapshotName))
-                {
-                    settingsEnumerable = client.GetConfigurationSettingsAsync(
-                        new SettingSelector
+                    await CallWithRequestTracing(async () =>
+                    {
+                        AsyncPageable<ConfigurationSetting> pageableSettings = client.GetConfigurationSettingsAsync(selector, cancellationToken);
+
+                        await foreach (Page<ConfigurationSetting> page in _options.PageableManager.GetPages(pageableSettings).ConfigureAwait(false))
                         {
-                            KeyFilter = loadOption.KeyFilter,
-                            LabelFilter = loadOption.LabelFilter
-                        },
-                        cancellationToken);
+                            // Server etag will never be null or empty because non-conditional List key-values API always returns 200
+                            ETag serverEtag = (ETag)page?.GetRawResponse()?.Headers.ETag;
+
+                            if (serverEtag == null || page.Values == null)
+                            {
+                                throw new RequestFailedException(ErrorMessages.InvalidConfigurationSettingPage);
+                            }
+
+                            foreach (ConfigurationSetting setting in page.Values)
+                            {
+                                data[setting.Key] = setting;
+                            }
+
+                            matchConditions.Add(new MatchConditions { IfNoneMatch = serverEtag });
+                        }
+                    }).ConfigureAwait(false);
+
+                    if (loadOption.IsFeatureFlagSelector)
+                    {
+                        watchedFfEtags[loadOption] = matchConditions;
```

Comment:
Any etag we store is for watching purposes right? If so, then just `ffEtags` should suffice.